rust/otel-arrow-rust/src/encode/record/array/string.rs

```diff
@@ -0,0 +1,179 @@
+use std::sync::Arc;
+
+use arrow::array::{
+    Array, ArrowPrimitiveType, DictionaryArray, StringArray, StringBuilder, StringDictionaryBuilder,
+};
+use arrow::datatypes::{ArrowDictionaryKeyType, DataType, UInt8Type, UInt16Type};
+use arrow::error::ArrowError;
+
+use super::dictionary::{DictionaryArrayBuilder, DictionaryBuilderError as Error, Result};
+use super::{ArrayBuilder, ArrayBuilderConstructor, dictionary::UpdateDictionaryIndexInto};
+
+impl ArrayBuilderConstructor for StringBuilder {
+    fn new() -> Self {
+        Self::new()
+    }
+}
+
+impl ArrayBuilder for StringBuilder {
+    type Native = String;
+
+    fn append_value(&mut self, value: &Self::Native) {
+        self.append_value(value);
+    }
+
+    fn finish(&mut self) -> super::ArrayWithType {
+        return super::ArrayWithType {
+            data_type: DataType::Utf8,
+            array: Arc::new(self.finish()),
+        };
+    }
+}
+
+impl<T> ArrayBuilderConstructor for StringDictionaryBuilder<T>
+where
+    T: ArrowDictionaryKeyType,
+{
+    fn new() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> DictionaryArrayBuilder for StringDictionaryBuilder<T>
+where
+    T: ArrowDictionaryKeyType + ArrowPrimitiveType,
+    <T as ArrowPrimitiveType>::Native: Into<usize>,
+{
+    type Native = String;
+
+    fn append_value(&mut self, value: &Self::Native) -> Result<usize> {
+        match self.append(value) {
+            Ok(index) => Ok(index.into()),
+            Err(ArrowError::DictionaryKeyOverflowError) => Err(Error::DictOverflow {}),
+
+            // safety: shouldn't happen. The only error type append should
+            // return should be for dictionary overflows
+            Err(e) => panic!("{}", e),
+        }
+    }
+
+    fn finish(&mut self) -> super::ArrayWithType {
+        return super::ArrayWithType {
+            data_type: DataType::Dictionary(Box::new(T::DATA_TYPE), Box::new(DataType::Utf8)),
+            array: Arc::new(self.finish()),
+        };
+    }
+}
+
+impl<K> super::dictionary::ConvertToNativeHelper for StringDictionaryBuilder<K>
+where
+    K: ArrowDictionaryKeyType,
+{
+    type Accessor = StringArray;
+}
+
+impl UpdateDictionaryIndexInto<StringDictionaryBuilder<UInt16Type>>
+    for StringDictionaryBuilder<UInt8Type>
+{
+    fn upgrade_into(&mut self) -> StringDictionaryBuilder<UInt16Type> {
+        // TODO there should be an optimized way to implement this. Thinking we could
+        // create a new builder with the same keys (but use `cast` kernel) cast them
+        // to u16 then reuse the same values
```

Comment:
That will be a great optimization!

Reply:
I'm not sure it's currently possible to covert the builders directly like this without copying the values. I'm going to leave this as a TODO for now and start a thread in the arrow-rs discord to get clarity. Something tells me there might be some changes required in arrow-rs to enable this type of thing

Reply:
Created a followup issue for this https://github.com/open-telemetry/otel-arrow/issues/536

Reply:
Related PR to add efficient method to builder to upgrade the key type https://github.com/apache/arrow-rs/pull/7611